File: rust/var-core/src/value.rs
//! The dynamic value model — the Rust replacement for Java var-core's `Object`
//! with `instanceof Map`/`List`/`String` duck-typing (see `CellDiff.java`,
//! `DocStringDiff.java`, `ParamDiff.java`). One closed enum carries handler
//! arguments, handler returns, thread-through state, row objects, table rows,
//! and the conformance wire values.
//!
//! Equality is derived `PartialEq`, the analog of Java's `Objects.equals`:
//! `Int(2) != Float(2.0)` (Java `Integer(2).equals(Double(2.0))` is false), and
//! `Map` equality is order-insensitive (`BTreeMap`), matching `Map.of(...)`
//! vs `LinkedHashMap` equality in the Java tests.

use std::collections::BTreeMap;

/// A dynamic JSON-ish value. `BTreeMap` gives order-insensitive map equality and
/// a free recursive key-sort for canonical JSON.
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    /// Integer (Java `Integer`/`Long`; `{int}` transforms here).
    Int(i64),
    /// Floating-point (Java `Double`); serialized as an integer when integral.
    Float(f64),
    String(String),
    List(Vec<Value>),
    Map(BTreeMap<String, Value>),
}

impl Value {
    /// A short type name (for `ReturnShapeError` messages, mirroring Java's
    /// `getClass().getSimpleName()`).
    pub fn type_name(&self) -> &'static str {
        match self {
            Value::Null => "null",
            Value::Bool(_) => "Boolean",
            Value::Int(_) => "Integer",
            Value::Float(_) => "Double",
            Value::String(_) => "String",
            Value::List(_) => "List",
            Value::Map(_) => "Map",
        }
    }

    /// Builds a [`Value::List`] from anything iterable of `Value`.
    pub fn list(items: impl IntoIterator<Item = Value>) -> Value {
        Value::List(items.into_iter().collect())
    }

    /// Builds a [`Value::Map`] from `(String, Value)` pairs.
    pub fn map(entries: impl IntoIterator<Item = (String, Value)>) -> Value {
        Value::Map(entries.into_iter().collect())
    }
}

impl From<i64> for Value {
    fn from(v: i64) -> Value {
        Value::Int(v)
    }
}

impl From<i32> for Value {
    fn from(v: i32) -> Value {
        Value::Int(i64::from(v))
    }
}

impl From<bool> for Value {
    fn from(v: bool) -> Value {
        Value::Bool(v)
    }
}

impl From<f64> for Value {
    fn from(v: f64) -> Value {
        Value::Float(v)
    }
}

impl From<&str> for Value {
    fn from(v: &str) -> Value {
        Value::String(v.to_string())
    }
}

impl From<String> for Value {
    fn from(v: String) -> Value {
        Value::String(v)
    }
}

impl From<Vec<Value>> for Value {
    fn from(v: Vec<Value>) -> Value {
        Value::List(v)
    }
}

File: go/core/value.go
package varcore

// The dynamic value model — Go's replacement for the TS/JS raw values and the
// Rust `Value` enum. One tagged struct carries handler arguments, handler
// returns, thread-through state, row objects, table rows, and the conformance
// wire values.
//
// Equality (ValueEqual) mirrors Rust's derived PartialEq / Java's
// Objects.equals: Int(2) != Float(2.0), and Map equality is order-insensitive.

// ValueKind is the discriminant of a Value.
type ValueKind int

const (
	KindNull ValueKind = iota
	KindBool
	KindInt
	KindFloat
	KindString
	KindList
	KindMap
)

// Value is a dynamic JSON-ish value.
type Value struct {
	Kind  ValueKind
	Bool  bool
	Int   int64
	Float float64
	Str   string
	List  []Value
	Map   map[string]Value
}

// NullValue is the null value.
var NullValue = Value{Kind: KindNull}

// BoolValue builds a boolean Value.
func BoolValue(b bool) Value { return Value{Kind: KindBool, Bool: b} }

// IntValue builds an integer Value ({int} transforms here).
func IntValue(i int64) Value { return Value{Kind: KindInt, Int: i} }

// FloatValue builds a floating-point Value; serialized as an integer when integral.
func FloatValue(f float64) Value { return Value{Kind: KindFloat, Float: f} }

// StrValue builds a string Value.
func StrValue(s string) Value { return Value{Kind: KindString, Str: s} }

// ListValue builds a list Value.
func ListValue(items ...Value) Value {
	if items == nil {
		items = []Value{}
	}
	return Value{Kind: KindList, List: items}
}

// ListOf builds a list Value from a slice.
func ListOf(items []Value) Value {
	if items == nil {
		items = []Value{}
	}
	return Value{Kind: KindList, List: items}
}

// MapValue builds a map Value.
func MapValue(m map[string]Value) Value {
	if m == nil {
		m = map[string]Value{}
	}
	return Value{Kind: KindMap, Map: m}
}

// TypeName is a short type name for ReturnShapeError messages, mirroring
// the reference's getClass().getSimpleName().
func (v Value) TypeName() string {
	switch v.Kind {
	case KindNull:
		return "null"
	case KindBool:
		return "Boolean"
	case KindInt:
		return "Integer"
	case KindFloat:
		return "Double"
	case KindString:
		return "String"
	case KindList:
		return "List"
	case KindMap:
		return "Map"
	}
	return "null"
}

// IsNull reports whether v is null (or the zero Value).
func (v Value) IsNull() bool { return v.Kind == KindNull }

// AsInt returns the integer value and true if v is an Int.
func (v Value) AsInt() (int64, bool) {
	if v.Kind == KindInt {
		return v.Int, true
	}
	return 0, false
}

// AsFloat returns the float value and true if v is a Float.
func (v Value) AsFloat() (float64, bool) {
	if v.Kind == KindFloat {
		return v.Float, true
	}
	return 0, false
}

// AsBool returns the boolean value and true if v is a Bool.
func (v Value) AsBool() (bool, bool) {
	if v.Kind == KindBool {
		return v.Bool, true
	}
	return false, false
}

// AsString returns the string value and true if v is a String.
func (v Value) AsString() (string, bool) {
	if v.Kind == KindString {
		return v.Str, true
	}
	return "", false
}

// AsList returns the list and true if v is a List.
func (v Value) AsList() ([]Value, bool) {
	if v.Kind == KindList {
		return v.List, true
	}
	return nil, false
}

// AsMap returns the map and true if v is a Map.
func (v Value) AsMap() (map[string]Value, bool) {
	if v.Kind == KindMap {
		return v.Map, true
	}
	return nil, false
}

// ValueEqual reports structural equality (Rust PartialEq / Objects.equals):
// distinct kinds are never equal (so Int(2) != Float(2.0)), lists compare
// order-sensitively, maps order-insensitively.
func ValueEqual(a, b Value) bool {
	if a.Kind != b.Kind {
		return false
	}
	switch a.Kind {
	case KindNull:
		return true
	case KindBool:
		return a.Bool == b.Bool
	case KindInt:
		return a.Int == b.Int
	case KindFloat:
		return a.Float == b.Float
	case KindString:
		return a.Str == b.Str
	case KindList:
		if len(a.List) != len(b.List) {
			return false
		}
		for i := range a.List {
			if !ValueEqual(a.List[i], b.List[i]) {
				return false
			}
		}
		return true
	case KindMap:
		if len(a.Map) != len(b.Map) {
			return false
		}
		for k, av := range a.Map {
			bv, ok := b.Map[k]
			if !ok || !ValueEqual(av, bv) {
				return false
			}
		}
		return true
	}
	return false
}
